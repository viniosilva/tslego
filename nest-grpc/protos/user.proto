syntax = "proto3";

package user;

service UserService {
  rpc getUsers (Empty) returns (UsersResponse) {}
  rpc getUserById (UserRequestId) returns (UserResponse) {}
  rpc createUser (UserRequest) returns (UserResponse) {}
  rpc updateUserById (User) returns (UserResponse) {}
  rpc removeUserById (UserRequestId) returns (MessageRequest) {}
}

message Empty {}

message User {
  int32 id = 1;
  string name = 2;
}

message UserRequestId {
  int32 id = 1;
}

message UserRequest {
  string name = 1;
}

message UserResponse {
  User data = 1;
}

message UsersResponse {
  repeated User data = 1;
}

message MessageRequest {
  string message = 1;
}
