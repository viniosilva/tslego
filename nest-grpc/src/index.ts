import { NestFactory } from '@nestjs/core';
import { MicroserviceOptions, Transport } from '@nestjs/microservices';
import { AppModule } from './AppModule';
import { apiConfig } from './common/config';
import { logger } from './common/logger';

async function bootstrap() {
  const url = `${apiConfig.host}:${apiConfig.port}`;
  const app = await NestFactory.createMicroservice<MicroserviceOptions>(
    AppModule,
    {
      logger: false,
      transport: Transport.GRPC,
      options: {
        package: 'user',
        protoPath: 'protos/user.proto',
        url,
      },
    },
  );

  app.listen(() => logger.info(`App listening at [${url}]`));
}

bootstrap();
