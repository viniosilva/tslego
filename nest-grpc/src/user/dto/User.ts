import { IsNotEmpty, IsNumber, IsString, Length } from 'class-validator';

export class User {
  @IsNumber()
  @IsNotEmpty()
  id: number;

  @IsString()
  @IsNotEmpty()
  @Length(3, 100)
  name: string;
}

export class UserRequest {
  @IsString()
  @IsNotEmpty()
  @Length(3, 100)
  name: string;
}

export class UserRequestId {
  @IsNumber()
  @IsNotEmpty()
  id: number;
}
