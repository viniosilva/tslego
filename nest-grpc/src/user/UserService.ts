import { BadRequestException } from '@nestjs/common';
import { validateSync } from 'class-validator';
import { User, UserRequest, UserRequestId } from './dto/User';
import { NotFoundException } from '../common/exception/NotFoundException';

export class UserService {
  private readonly users: User[] = [];

  validateUser(data: User) {
    const user = new User();
    user.id = data.id;
    user.name = data.name;

    this.validate(user);
  }

  validateUserRequest(data: UserRequest) {
    const user = new UserRequest();
    user.name = data.name;

    this.validate(user);
  }

  validateUserRequestId(data: UserRequestId) {
    const user = new UserRequestId();
    user.id = data.id;

    this.validate(user);
  }

  private validate(object: any) {
    const errors = validateSync(object);
    if (errors.length > 0) {
      throw new BadRequestException(errors.join('; '));
    }
  }

  getUsers(): User[] {
    return this.users;
  }

  getUserById(userId: number): User {
    const user = this.users.find(({ id }) => id === userId);
    if (!user) throw new NotFoundException('User not found');
    return user;
  }

  createUser(userRequest: UserRequest): User {
    const user = new User();
    user.id = this.users.length + 1;
    user.name = userRequest.name;

    this.users.push(user);
    return user;
  }

  updateUserById(userId: number, userRequest: UserRequest): User {
    const user = this.getUserById(userId);
    user.name = userRequest.name;
    return user;
  }

  removeUserById(userId: number): void {
    let indexOf: number;
    this.users.some((user, index) => {
      if (user.id === userId) {
        indexOf = index;
        return true;
      }
    });

    if (!isNaN(indexOf)) this.users.splice(indexOf, 1);
  }
}
