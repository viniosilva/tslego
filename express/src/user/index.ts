import { UserController } from './UserController';
import { UserService } from './UserService';

const userService = new UserService();
export const userController = new UserController(userService);
