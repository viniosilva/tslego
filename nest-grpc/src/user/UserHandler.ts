import { Controller, Inject } from '@nestjs/common';
import { GrpcMethod } from '@nestjs/microservices';
import { Metadata, ServerUnaryCall } from 'grpc';
import { Response } from 'src/dto/Response';
import { User, UserRequest, UserRequestId } from './dto/User';
import { UserService } from './UserService';

@Controller()
export class UserHandler {
  constructor(
    private readonly userService: UserService,
    @Inject('ErrorHandler') private readonly errorHandler: Function,
  ) {}

  @GrpcMethod('UserService', 'getUsers')
  getUsers(
    _data: any,
    _metadata: Metadata,
    _call: ServerUnaryCall<Response>,
  ): Response {
    try {
      const users = this.userService.getUsers();
      return { data: users };
    } catch (err) {
      this.errorHandler(err);
    }
  }

  @GrpcMethod('UserService', 'getUserById')
  getUserById(
    data: UserRequestId,
    _metadata: Metadata,
    _call: ServerUnaryCall<Response>,
  ): Response {
    try {
      this.userService.validateUserRequestId(data);

      const user = this.userService.getUserById(data.id);
      return { data: user };
    } catch (err) {
      this.errorHandler(err);
    }
  }

  @GrpcMethod('UserService', 'createUser')
  createUser(
    data: UserRequest,
    _metadata: Metadata,
    _call: ServerUnaryCall<Response>,
  ): Response {
    try {
      this.userService.validateUserRequest(data);

      const user = this.userService.createUser(data);
      return { data: user };
    } catch (err) {
      this.errorHandler(err);
    }
  }

  @GrpcMethod('UserService', 'updateUserById')
  updateUserById(
    data: User,
    _metadata: Metadata,
    _call: ServerUnaryCall<Response>,
  ): Response {
    try {
      this.userService.validateUser(data);

      const user = this.userService.updateUserById(data.id, data);
      return { data: user };
    } catch (err) {
      this.errorHandler(err);
    }
  }

  @GrpcMethod('UserService', 'removeUserById')
  removeUserById(
    data: UserRequestId,
    _metadata: Metadata,
    _call: ServerUnaryCall<Response>,
  ): Response {
    try {
      this.userService.validateUserRequestId(data);

      this.userService.removeUserById(data.id);
      return { message: 'OK' };
    } catch (err) {
      this.errorHandler(err);
    }
  }
}
