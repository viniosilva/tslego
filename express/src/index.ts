import { apiConfig } from './common/config';
import { logger } from './common/logger';
import { Server } from './server';
import { userController } from './user';

const routes = [userController.router];
const server = new Server(
  apiConfig.host,
  apiConfig.port,
  logger,
  apiConfig.swaggerConfig,
  routes,
);

server.start();
