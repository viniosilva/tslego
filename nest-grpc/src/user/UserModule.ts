import { Module, Provider } from '@nestjs/common';
import { errorHandler } from '../common/errorHandler';
import { UserHandler } from './UserHandler';
import { UserService } from './UserService';

const userProvider: Provider = {
  provide: UserService,
  useFactory: () => new UserService(),
};

const errorHandlerProvider: Provider = {
  provide: 'ErrorHandler',
  useValue: errorHandler
};

@Module({
  controllers: [UserHandler],
  providers: [userProvider, errorHandlerProvider],
})
export class UserModule {}
