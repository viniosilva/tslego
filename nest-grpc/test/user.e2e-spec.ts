import { Test, TestingModule } from '@nestjs/testing';
import { AppModule } from '../src/AppModule';
import { Response } from '../src/dto/Response';
import { UserHandler } from '../src/user/UserHandler';
import { Metadata, ServerUnaryCall } from 'grpc';

describe('AppController (e2e)', () => {
  const metadata = new Metadata();
  const call = {} as ServerUnaryCall<Response>;
  let userHandler: UserHandler;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    userHandler = moduleFixture.get<UserHandler>(UserHandler);
  });

  describe('getUsers', () => {
    it('should return empty user list', () => {
      const res = userHandler.getUsers({}, metadata, call);
      expect(res).toMatchSnapshot();
    });

    it('should return user list', async () => {
      userHandler.createUser({ name: 'TEST' }, metadata, call);

      const res = userHandler.getUsers({}, metadata, call);
      expect(res).toMatchSnapshot();
    });
  });

  describe('getUserById', () => {
    it('should return user', async () => {
      const { data } = userHandler.createUser({ name: 'TEST' }, metadata, call);

      const res = userHandler.getUserById({ id: data.id }, metadata, call);
      expect(res).toMatchSnapshot();
    });

    it('should throw not found exception', () => {
      const fn = () => userHandler.getUserById({ id: 0 }, metadata, call);
      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('createUser', () => {
    it('should return new user', () => {
      const res = userHandler.createUser({ name: 'TEST' }, metadata, call);
      expect(res).toMatchSnapshot();
    });

    it('should throw bad request exception', () => {
      const fn = () => userHandler.createUser({ name: 'T' }, metadata, call);
      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('updateUserById', () => {
    it('should return updated user', async () => {
      const { data } = userHandler.createUser({ name: 'TEST' }, metadata, call);

      const res = userHandler.updateUserById(
        { id: data.id, name: 'UPDATE TEST' },
        metadata,
        call,
      );

      expect(res).toMatchSnapshot();
    });

    it('should throw bad request exception', () => {
      const { data } = userHandler.createUser({ name: 'TEST' }, metadata, call);

      const fn = () =>
        userHandler.updateUserById({ id: data.id, name: 'T' }, metadata, call);

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw not found exception', () => {
      const fn = () =>
        userHandler.updateUserById({ id: 0, name: 'TEST' }, metadata, call);

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('removeUserById', () => {
    it('should remove user', async () => {
      const { data } = userHandler.createUser({ name: 'TEST' }, metadata, call);
      const res = userHandler.removeUserById({ id: data.id }, metadata, call);

      expect(res).toMatchSnapshot();
    });
  });
});
