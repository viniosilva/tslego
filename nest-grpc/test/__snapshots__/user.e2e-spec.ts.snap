// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) createUser should return new user 1`] = `
Object {
  "data": User {
    "id": 1,
    "name": "TEST",
  },
}
`;

exports[`AppController (e2e) createUser should throw bad request exception 1`] = `"Internal server error"`;

exports[`AppController (e2e) getUserById should return user 1`] = `
Object {
  "data": User {
    "id": 1,
    "name": "TEST",
  },
}
`;

exports[`AppController (e2e) getUserById should throw not found exception 1`] = `"User not found"`;

exports[`AppController (e2e) getUsers should return empty user list 1`] = `
Object {
  "data": Array [],
}
`;

exports[`AppController (e2e) getUsers should return user list 1`] = `
Object {
  "data": Array [
    User {
      "id": 1,
      "name": "TEST",
    },
  ],
}
`;

exports[`AppController (e2e) removeUserById should remove user 1`] = `
Object {
  "message": "OK",
}
`;

exports[`AppController (e2e) updateUserById should return updated user 1`] = `
Object {
  "data": User {
    "id": 1,
    "name": "UPDATE TEST",
  },
}
`;

exports[`AppController (e2e) updateUserById should throw bad request exception 1`] = `"Internal server error"`;

exports[`AppController (e2e) updateUserById should throw not found exception 1`] = `"User not found"`;
